File: BinaryTree/js/treeLevels.js
const treeLevels = (root) => {
  if (root === null) return [];

  const levels = [];
  const queue = [{ node: root, levelNum: 0 }];
  while (queue.length > 0) {
    const { node, levelNum } = queue.shift();

    if (levels.length === levelNum) {
      levels[levelNum] = [node.val];
    } else {
      levels[levelNum].push(node.val);
    }

    if (node.left !== null) queue.push({ node: node.left, levelNum: levelNum + 1 });
    if (node.right !== null) queue.push({ node: node.right, levelNum: levelNum + 1 });
  }

  return levels;
};


const treeLevels = (root) => {
  const levels = [];
  _treeLevels(root, levels, 0);
  return levels;
};

const _treeLevels = (root, levels, levelNum) => {
  if (root === null) return;

  if (levels.length === levelNum) {
    levels[levelNum] = [root.val];
  } else {
    levels[levelNum].push(root.val);
  }

  _treeLevels(root.left, levels, levelNum + 1);
  _treeLevels(root.right, levels, levelNum + 1);
};

File: BinaryTree/go/treeLevels.go
package binarytree

// TreeLevels groups values by their depth starting at 0.
func TreeLevels(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}

	var levels [][]int
	type entry struct {
		node  *TreeNode
		level int
	}
	queue := []entry{{node: root, level: 0}}

	for len(queue) > 0 {
		item := queue[0]
		queue = queue[1:]

		if len(levels) == item.level {
			levels = append(levels, []int{})
		}
		levels[item.level] = append(levels[item.level], item.node.Val)

		if item.node.Left != nil {
			queue = append(queue, entry{node: item.node.Left, level: item.level + 1})
		}
		if item.node.Right != nil {
			queue = append(queue, entry{node: item.node.Right, level: item.level + 1})
		}
	}

	return levels
}
